### ISSUE TITLE
Vue | Carousel issue: Regression, v-show breaks the carousel

### ISSUE THREAD

**Issue by pzanitti** [OnsenUI/OnsenUI#2278]
__Environment__

```
[Core]
  onsenui latest
  
[Framework]
  vue 2.5.9

[Framework binding]
 vue-onsenui latest

[Platform]
  Desktop - OS X 10.13.1 64-bit

[Browser]
  Desktop - Chrome 63.0.3239.84 on OS X 10.13.1 64-bit
```

__Encountered problem__
When migrating to the latest version, using a v-show attribute on a v-ons-carousel breaks its behavior.

Can also reproduce on onsenui 2.7.2.

This worked on version:
"onsenui": "~2.5.1",
"vue": "^2.4.0",
"vue-onsenui": "~2.1.0",

__How to reproduce__
Take the v-ons-carousel example, put a v-show on it and some way to toggle its condition. When it is toggled on, the carousel will show but its behavior will be broken.

  
  [__Demo link__](https://tutorial.onsen.io/?issue)

  

- __HTML__

```html
<template id="main">
  <v-ons-page>
    <v-ons-toolbar>
      <div class="left"><ons-toolbar-button @click="carouselIndex > 0 && carouselIndex--">i--</ons-toolbar-button></div>
      <div class="center">
        <span v-if="carouselEnabled">Index: {{carouselIndex}}</span>
        <span v-else @click="carouselEnabled = true">Enable</span>
      </div>
      <div class="right"><ons-toolbar-button @click="carouselIndex < 2 && carouselIndex++">i++</ons-toolbar-button></div>
    </v-ons-toolbar>

    <v-ons-carousel fullscreen swipeable auto-scroll overscrollable
      :index.sync="carouselIndex"
      v-show="carouselEnabled"
    >
      <v-ons-carousel-item v-for="(value, key) in items" :style="{backgroundColor: value}">
        <div style="text-align: center; font-size: 30px; margin-top: 20px; color: #fff;">{{key}}</div>
      </v-ons-carousel-item>
    </v-ons-carousel>

    <div :style="dots">
      <span :index="dotIndex - 1" v-for="dotIndex in Object.keys(items).length" :key="dotIndex" style="cursor: pointer" @click="carouselIndex = dotIndex - 1">
        {{ carouselIndex === dotIndex - 1 ? '\u25CF' : '\u25CB' }}
      </span>
    </div>
  </v-ons-page>
</template>

<div id="app"></div>
```

- __JS__

```javascript
new Vue({
  el: '#app',
  template: '#main',
  data() {
    return {
      carouselEnabled: false,
      carouselIndex: 0,
      items: {
        BLUE: '#085078',
        DARK: '#373B44',
        ORANGE: '#D38312'
      },
      dots: {
        textAlign: 'center',
        fontSize: '30px',
        color: '#fff',
        position: 'absolute',
        bottom: '40px',
        left: 0,
        right: 0
      }
    };
  }
});
```

**Comment by frandiox** [comment]
@pzanitti Thanks for reporting! The carousel implementation has changed a lot recently to improve performance and to share code with the new swiping tabbar.

The reason of this issue is that the carousel needs to measure some stuff during its first rendering: item size, initial scroll, etc. If the carousel is hidden with `display:none` then all of these values will be `0` or `NaN`.

The commit where I closed this issue (d7a2b03) doesn't actually fix this completely but prevents error state. After displaying the carousel again, you will need to run something like `document.querySelector('ons-carousel').refresh()` to fix the carousel. I believe we cannot do this automatically because of Custom Elements API limitations. Vue is toggling `element.style.display` DOM property with `v-show` and we have no way of knowing when this happens. If they used `element.setAttribute('style', ...)` then we could use `attributeChangedCallback` but, unfortunately, this is not possible with DOM properties (unless we do dirty property override).

However, there's another solution. If you use `v-if` instead of `v-show`, the carousel will be able to setup properly on its first rendering. In order to use `v-if`, you need to wrap you content in `<div class="content">...</div>` (more about this [here](https://onsen.io/v2/guide/vue/#vonspage-compilation)).

```
<v-ons-page>
  <v-ons-toolbar />

  <div class="content">
    <v-ons-carousel v-if="carouselEnabled">...</v-ons-carousel>
  </div>
</v-ons-page>
```

If you need to toggle it afterwards, then you can use a `v-show` normally. Just avoid it during the first rendering (or call the element's `refresh` method).

**Comment by frandiox** [comment]
For the record, I tried something like this 07f859b6e76ab75732c3898a20647dccb136b3f3 and works in Chrome and Firefox. However, style properties are not configurable in Safari and crashes :disappointed:
